### PR TITLE
Fix static tls

### PR DESCRIPTION
**Pull request by agorshk**

### CODE REVIEW

**Comment by agorshk** [comment]
nightly http://intel-ci.intel.com/eb5b12e0-d471-f1b9-96dc-8cdcd4b723a1

File: cpp/daal/src/threading/threading.h [hunk 1]
```diff
@@ -337,14 +352,23 @@ class static_tls
...
-        const void * ac = static_cast<const void *>(&lambda);
+        lambdaType * locall = new lambdaType(lambda);
+        _deleter            = new static_tls_deleter_<lambdaType>();
+
+        const void * ac = static_cast<const void *>(locall);
         void * a        = const_cast<void *>(ac);
         _creater        = a;
 
         _creater_func = creater_func<F, lambdaType>;
+        int abc       = 0;
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/threading/threading.h:363]: For what?

File: cpp/daal/src/threading/threading.h [hunk 2]
```diff
@@ -337,14 +352,23 @@ class static_tls
             _storage[i] = nullptr;
         }
 
-        const void * ac = static_cast<const void *>(&lambda);
+        lambdaType * locall = new lambdaType(lambda);
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/threading/threading.h:355]: You do "new", but I don't see "delete" for this.

File: cpp/daal/src/threading/threading.h [hunk 3]
```diff
@@ -337,14 +352,23 @@ class static_tls
...
+        _deleter            = new static_tls_deleter_<lambdaType>();
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/threading/threading.h:356]: Now, ```locall``` and ```_deleter``` can be ```nullptr``` due to potential memory allocation fail.
Need to handle this somehow.

File: cpp/daal/src/threading/threading.h (hunk 1 above)
Comment (agorshk) [reply cpp/daal/src/threading/threading.h]: it was removed, it used to debugging

File: cpp/daal/src/threading/threading.h (hunk 3 above)
Comment (agorshk) [reply cpp/daal/src/threading/threading.h:356]: it was fixed

**Comment by agorshk** [comment]
nightly (icc_clang) http://intel-ci.intel.com/eb6146f1-bb22-f102-8465-8cdcd4b723a1

**Review by SmirnovEgorRu — APPROVED** [review]

**Pull request merged**